skyrl-tx/tx/layers/lora.py

```diff
@@ -0,0 +1,106 @@
+from flax import nnx
+import jax
+from jax import numpy as jnp
+
+
+def Param(*shape: int, dtype: jnp.dtype, kernel_init: nnx.Initializer, rngs: nnx.Rngs):
+    return nnx.Param(kernel_init(rngs.param(), shape, dtype))
+
+
+class LoRAMixin:
+    """A mixin for flax NNX modules to add multi-adapter LoRA support.
+    This mixin adds LoRA parameters (lora_A, lora_B) and methods to apply
+    the low-rank adaptation to a base module's output. It is designed to
+    be used with layers like nnx.Linear.
+    """
+
+    lora_scaling: nnx.Param | None
+    lora_A: nnx.Param | None
+    lora_B: nnx.Param | None
+
+    def init_lora(
+        self, *, max_lora_adapters: int, in_features: int, out_features: int,
+        max_lora_rank: int, sharding: jax.sharding.PartitionSpec,
+        dtype: jnp.dtype, rngs: nnx.Rngs,
+    ) -> None:
+        self.in_features = in_features
+        self.out_features = out_features
+        self.max_lora_adapters = max_lora_adapters
+        self.max_lora_rank = max_lora_rank
+
+        if max_lora_adapters == 0:
+            self.lora_scaling = None
+            self.lora_A = None
+            self.lora_B = None
+        else:
+            self.lora_scaling = Param(
+                max_lora_adapters, dtype=dtype,
+                kernel_init=nnx.initializers.constant(1.0), rngs=rngs,
+            )
+            self.lora_A = Param(
+                max_lora_adapters, in_features, max_lora_rank, dtype=dtype,
+                kernel_init=nnx.with_partitioning(
+                    nnx.initializers.he_uniform(),
```

Comment:
It is worth checking this. The blog https://thinkingmachines.ai/blog/lora/ mentions they use a uniform distribution for A with scale 1 / sqrt(in_features) and there is also https://github.com/microsoft/LoRA/blob/a0a92e0f26c067cf94747bdbf1ce73793fa44d19/loralib/layers.py#L124

I believe this should be right and the same as pytorch though, see https://discuss.pytorch.org/t/clarity-on-default-initialization-in-pytorch/84696 (namely a = sqrt(5) corresponds to uniform distribution with scale 1 / sqrt(in_features), corresponds to he_uniform).